client/mm/balance_handler.go

```diff
@@ -0,0 +1,949 @@
+	defer feed.ReturnFeed()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case n := <-feed.C:
+			b.handleNotification(n)
+		}
+	}
+}
+
+func newBalanceHandler(cfgs []*BotConfig, core clientCore, log dex.Logger) (*balanceHandler, error) {
+	bh := &balanceHandler{
+		botBalances: make(map[string]*botBalance),
+		orders:      make(map[order.OrderID]*orderInfo),
+		core:        core,
+		log:         log,
+	}
+
+	type trackedBalance struct {
+		balanceAvailable uint64
+		balanceReserved  uint64
+	}
+
+	balanceTracker := make(map[uint32]*trackedBalance)
+
+	for _, cfg := range cfgs {
+		if _, found := balanceTracker[cfg.BaseAsset]; !found {
+			bal, err := core.AssetBalance(cfg.BaseAsset)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get balance for asset %d: %v", cfg.BaseAsset, err)
+			}
+			balanceTracker[cfg.BaseAsset] = &trackedBalance{
+				balanceAvailable: bal.Available,
+			}
+		}
+
+		if _, found := balanceTracker[cfg.QuoteAsset]; !found {
+			bal, err := core.AssetBalance(cfg.QuoteAsset)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get balance for asset %d: %v", cfg.QuoteAsset, err)
+			}
+			balanceTracker[cfg.QuoteAsset] = &trackedBalance{
+				balanceAvailable: bal.Available,
+			}
+		}
+
+		baseBalance := balanceTracker[cfg.BaseAsset]
+		quoteBalance := balanceTracker[cfg.QuoteAsset]
+
+		var baseRequired, quoteRequired uint64
+		if cfg.BaseBalanceType == Percentage {
+			baseRequired = baseBalance.balanceAvailable * cfg.BaseBalance / 100
+		} else {
+			baseRequired = cfg.BaseBalance
+		}
+
+		if cfg.QuoteBalanceType == Percentage {
+			quoteRequired = quoteBalance.balanceAvailable * cfg.QuoteBalance / 100
+		} else {
+			quoteRequired = cfg.QuoteBalance
+		}
+
+		if baseRequired == 0 && quoteRequired == 0 {
+			return nil, fmt.Errorf("both base and quote balance are zero for market %s-%d-%d",
+				cfg.Host, cfg.BaseAsset, cfg.QuoteAsset)
+		}
+
+		if baseRequired > baseBalance.balanceAvailable-baseBalance.balanceReserved {
+			return nil, fmt.Errorf("insufficient balance for asset %d", cfg.BaseAsset)
+		}
+		if quoteRequired > quoteBalance.balanceAvailable-quoteBalance.balanceReserved {
+			return nil, fmt.Errorf("insufficient balance for asset %d", cfg.QuoteAsset)
+		}
+
+		baseBalance.balanceReserved += baseRequired
+		quoteBalance.balanceReserved += quoteRequired
+
+		mktID := dexMarketID(cfg.Host, cfg.BaseAsset, cfg.QuoteAsset)
+		bh.botBalances[mktID] = &botBalance{
+			balances: map[uint32]uint64{
+				cfg.BaseAsset:  baseRequired,
+				cfg.QuoteAsset: quoteRequired,
+			},
+		}
+	}
+
+	return bh, nil
+}
+
+// coreWithSegregatedBalance implements the clientCore interface. A separate
+// instance should be created for each bot, and the core functions will behave
+// as if the entire balance of the wallet is the amount that has been reserved
+// for the bot.
+type coreWithSegregatedBalance struct {
+	balanceHandler *balanceHandler
+	botID          string
+	core           clientCore
```

Comment:
Named field is fine since it makes infinite recursion harder to do accidentally when shadowing a method like `Trade`, but it is possible to embed and remove all the dummy passthrough methods.  Did you have other considerations?

Reply:
Nope, embedding sounds good.